common/src/Model/Brush.cpp

```diff
@@ -1411,5 +1416,102 @@ namespace TrenchBroom {
             assert(intersects.hasResult());
             return intersects.result();
         }
+
+        void Brush::invalidateVertexCache() {
+            m_rendererCacheValid = false;
+            m_cachedVertices.clear();
+            m_cachedEdges.clear();
+            m_cachedFacesSortedByTexture.clear();
+        }
+
+        void Brush::validateVertexCache() const {
+            if (m_rendererCacheValid)
+                return;
+
+            ensure(m_geometry != nullptr, "geometry is null");
+
+            // build vertex cache and face cache
+
+            m_cachedVertices.clear();
+            m_cachedVertices.reserve(vertexCount());
+
+            m_cachedFacesSortedByTexture.clear();
+            m_cachedFacesSortedByTexture.reserve(faceCount());
+
+            for (BrushFace* face : m_faces) {
+                const size_t startIndex = m_cachedVertices.size();
+
+                const BrushHalfEdge *first = face->geometry()->boundary().front();
+                const BrushHalfEdge *current = first;
+                do {
+                    BrushVertex* vertex = current->origin();
+
+                    // Set the vertex payload to the index, relative to the brush's first vertex being 0.
+                    // This is used below when building the edge cache.
+                    // NOTE: we'll overwrite the payload as we visit the same vertex several times while visiting
+                    // different faces, this is fine.
+                    const size_t currentIndex = m_cachedVertices.size();
+                    vertex->setPayload(static_cast<GLuint>(currentIndex));
+
+                    const Vec3 &position = vertex->position();
```

Comment:
please use `const Vec3& position` -- I'd like to keep style consistent if possible.